pkg/distributor/distributor.go

```diff
@@ -339,6 +398,59 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	}
 }
 
+func min(x1, x2 int) int {
+	if x1 < x2 {
+		return x1
+	}
+	return x2
+}
+
+// shardStream shards (divides) the given stream into smaller streams.
+//
+// It will derive N streams (and its entries) from the given stream, where N is the sharding size for the given stream.
+func shardStream(stream logproto.Stream, cfg Config, streamSharder StreamSharder, userID string) ([]uint32, []streamTracker) {
+	logger := util_log.Logger
+	shards, ok := streamSharder.ShardsFor(stream)
+	if !ok || shards == 1 {
+		return []uint32{util.TokenFor(userID, stream.Labels)}, []streamTracker{{stream: stream}}
+	}
+
+	derivedKeys := make([]uint32, 0, shards)
+	derivedStreams := make([]streamTracker, 0, shards)
+
+	if cfg.ShardStreams.Debug {
+		level.Warn(logger).Log("msg", "sharding request with mode", "mode", cfg.ShardStreams.Mode)
+	}
+
+	entriesPerWindow := float64(len(stream.Entries)) / float64(shards) // divide and keep decimal value.
+	for i := 0; i < shards; i++ {
+		streamCopy := logproto.Stream{}
+		lbs, err := syntax.ParseLabels(stream.Labels)
+		if err != nil {
+			level.Error(logger).Log("msg", "couldn't extract labels from stream", "stream", streamCopy.Labels)
+			continue
+		}
+
+		lbs = append(lbs, labels.Label{Name: ShardLbName, Value: fmt.Sprintf("%d", i)})
```

Comment:
The `fmt.Sprintf` here may be quite expensive; I would suggest benchmarking this function as it's in the critical path.

Reply:
I pushed a benchmark and a change to not use `fmt.Sprintf` and you are right, not using `fmt.Sprintf` did help. There's other optimizations that we can do too, such as not parsing all labels just to append the shard label, but I prefer doing them in a follow-up PR to make reviewing easier.

Reply:
Works for me :+1: thanks!